python/tvm/script/_parser/core/doc.py

```diff
@@ -0,0 +1,361 @@
+# Licensed to the Apache Software Foundation (ASF) under one
+# or more contributor license agreements.  See the NOTICE file
+# distributed with this work for additional information
+# regarding copyright ownership.  The ASF licenses this file
+# to you under the Apache License, Version 2.0 (the
+# "License"); you may not use this file except in compliance
+# with the License.  You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+# pylint: disable=missing-docstring
+import ast
```

Comment:
This should be using [`synr`](https://github.com/octoml/synr) instead of the python ast module. Synr provides a stable ast across multiple versions of python, unlike `ast` which changes version to version. Synr is also used by the old tvmscript parser.

Reply:
We are using the new stable AST provided by TVMScript. Its stability has been tested across python 3.7 - 3.10 with 1000 tests in the testsuite.

Not using `synr` is intentional, and the reason is stated and widely discussed in the RFC: for metaprogramming. During metaprogramming, python interpreter is used to evaluate pieces of the python AST, but honest back-translation from synr to python AST is not possible because it drops necessary information, for example, `x += 1` is normalized into `x = x + 1`.

Therefore, TVMScript's stable AST is designed with this in mind: it is an honest python 3.8 AST without any change. Any other versions of the AST is canonicalized to python 3.8 AST, and the file allow it to be translated back.

Reply:
> During metaprogramming, python interpreter is used to evaluate pieces of the python AST, but honest back-translation from synr to python AST is not possible because it drops necessary information, for example, x += 1 is normalized into x = x + 1.

I'm a little unclear on why normalization is a problem. `x += 1` and `x = x + 1` are equivalent statements.

Why not fix synr instead? It avoids having to rewrite large amounts of code. Sorry, but ditching synr wasn't made clear in the RFC. I was assuming you were going to use synr to maintain compatibility with multiple versions of python.

Reply:
glad you mentioned the engineering effort of maintaining synr, and that's exactly why we need a stable AST that's exactly the same as the one as python 3.8 :-)

At the time of writing, there are 1725 lines of code in [synr folder](https://github.com/octoml/synr/tree/main/synr). There are 22 tests developed to test all the features for parsing across all python versions; Meanwhile, only 361 lines of code are needed to maintain the translation and back-translation in-between python ast and the stable ast in TVMScript, and it's tested against thousands of real-world cases as mentioned in my previous reply.

The key here is that the stable ast is exactly honest with no change at all, so many logics can be implemented with some python programming tricks, for example, `__getattr__`.

To summarize, with this approach, the AST is more succinct, more stable, more stringently tested and more maintainable.

> x += 1 and x = x + 1 are equivalent statements.

it could be equivalent with certain assumptions, but not in general case, for example, 
```python
a[f()] += 1
```
is not equivalent to 
```python
a[f()] = a[f()] + 1
```
if `f` has side effects, which is completely possible in metaprogramming, then we cannot make assumption.

Reply:
just a question here @junrushao 
> with 1000 real-world TVMScripts in the testsuite;

do these tests make asserts about the IR to validate that they roundtrip correctly and that the parsed IR matches the Python? I agree there are a great many TVMScript in the tvm repo, but I am wondering about the logic that they all serve to validate this PR. I see one test case in this PR, so my concern here is that we may silently introduce a bug because those 1000 tests are not intended to validate the TVMScript and therefore don't assert it is parsed correctly. If the new TVMScript parser changes, then we could in effect decrease the coverage of 1000s of tests that rely on TVMScript to assert various compiler behaviors.

A more convincing argument here is that you've hooked the TVMScript parser before and after this change, dumped the IR, and validated case-by-case that the IR matches. Is that what's being argued?

Reply:
> do these tests make asserts about the IR to validate that they roundtrip correctly and that the parsed IR matches the Python?

Yes, this is roundtrip tests

Reply:
could you point me at the test suite so i can make sure we're on the same page?

Reply:
https://github.com/junrushao/tvmscript-testsuite

Reply:
ok, i agree that does look like a reasonable test. is there any reason these aren't upstreamed?